pd/src/consensus/worker.rs

```diff
@@ -247,164 +241,235 @@ impl Worker {
+        // at the epoch boundary
 
-            let mut staking_token_supply = reader
-                .asset_lookup(*STAKING_TOKEN_ASSET_ID)
-                .await?
-                .map(|info| info.total_supply)
-                .unwrap();
+        // TODO: right now we are not writing the updated voting power from validator statuses
+        // back to tendermint, so that we can see how the statuses are computed without risking
+        // halting the testnet. in the future we want to add code here to send the next voting
+        // powers back to tendermint.
+        Ok(Default::default())
+    }
+
+    /// Process the state transitions for the end of an epoch.
+    async fn end_epoch(&mut self) -> Result<()> {
+        let reader = self.state.private_reader();
+
+        let pending_block = self
+            .pending_block
+            .as_mut()
+            .expect("pending block must be Some in EndBlock");
 
-            // steps (foreach validator):
-            // - get the total token supply for the validator's delegation tokens
-            // - process the updates to the token supply:
-            //   - collect all delegations occurring in previous epoch and apply them (adds to supply);
-            //   - collect all undelegations started in previous epoch and apply them (reduces supply);
-            // - feed the updated (current) token supply into current_rates.voting_power()
-            // - persist both the current voting power and the current supply
-            //
+        let height = pending_block
+            .height
+            .expect("height must already have been set");
+
+        // We've finished processing the last block of `epoch`, so we've
+        // crossed the epoch boundary, and (prev | current | next) are:
+        let prev_epoch = pending_block
+            .epoch
+            .clone()
+            .expect("epoch must already have been set");
+        let current_epoch = prev_epoch.next();
+        let next_epoch = current_epoch.next();
+
+        tracing::info!(
+            ?height,
+            ?prev_epoch,
+            ?current_epoch,
+            ?next_epoch,
+            "crossed epoch boundary, processing rate updates"
+        );
+        metrics::increment_counter!("epoch");
 
-            /// FIXME: set this less arbitrarily, and allow this to be set per-epoch
-            /// 3bps -> 11% return over 365 epochs, why not
-            const BASE_REWARD_RATE: u64 = 3_0000;
+        // Find all the validators which were *not* slashed in this block
+        let well_behaved_validators = pending_block
+            .validator_state_changes
+            .iter()
+            // THIS IS A LOAD-BEARING NEGATION: we want all validators which are *NOT* slashed
+            .filter(|p| !matches!(p.1, ValidatorState::Slashed))
+            .map(|p| p.0.clone()) // get the validator ID
+            .collect::<Vec<_>>();
+
+        // Process unbonding notes and nullifiers for this epoch
```

Comment:
This is where unbonding actually happens in the `PendingBlock`.

Notice that we select only from the well behaved validators, and we select all notes/nullifiers whose unbonding height is equal to or less than the height of this epoch-ending block. This naturally means that if the epoch duration or the unbonding period are changed, two things are true:

1. unbonding still happens always on epoch boundaries, but
2. a change to either of these parameters will delay the unbonding of notes delegated before the change is applied by at most the new epoch duration.